### PR TITLE
AE-1010: Gnosis-Safe Multisig Monitor Agent Template

### PR DESCRIPTION
**Pull request by echoAE**
This agent template monitors events emitted by Gnosis-Safe v1.0.0, v1.1.1, v1.2.0, or v1.3.0 MultiSig contracts.  The agent template also monitors the Ether and token balances of the contract for any changes.

The following are examples for testing against blockchain contracts and transactions:

Synthetix: protocolDAO Address: 0xEb3107117FEAd7de89Cd14D463D340A2E6917769
Version v1.3.0
```json
{
  "developerAbbreviation": "AE",
  "protocolName": "Synthetix",
  "protocolAbbreviation": "SYN",
  "gnosisSafe": {
    "address": "0xEb3107117FEAd7de89Cd14D463D340A2E6917769",
    "version": "v1.3.0"
  }
}
```

Example transactions for testing are:
ExecutionSuccess: 0x40ee75e7e9a4480fbad271be98f48ba24348e6332ab66a380a96ae637c6f60ff
ApproveHash: 0xc25232da7a7891184adb550360ed20419654584c6a5fa0dc23da076cecf774d0
ExecutionFailure: 0xda379176748082ce5b7dea58de0e247cb2487962264f9cd594f00eb4cfcc4af2

BitFinex 3 MultiSig Address: 0xC61b9BB3A7a0767E3179713f3A5c7a9aeDCE193C
Version: v1.1.1
```json
{
  "developerAbbreviation": "AE",
  "protocolName": "BitFinex",
  "protocolAbbreviation": "BITFI",
  "gnosisSafe": {
    "address": "0xC61b9BB3A7a0767E3179713f3A5c7a9aeDCE193C",
    "version": "v1.1.1"
  }
}
```

Example transactions for testing are:
ApproveHash Tx: 0x6a1b2bad39f9b9af38f90bb5cac4a75ff3ef092457e4a6cd15e18243d93c0e9c
ExecutionSuccess Tx: 0x5042f026c19a247a3300fb9ed0eca5ba57210793d5809818999b77e38ddd9e56

### CODE REVIEW

File: gnosis-safe-multisig/SETUP.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        gnosis_safe.json
+      ERC20.json
+```
+
+8. Install all related `npm` packages using `npm i`.  This will create a `package-lock.json` file alongside
+package.json.
+
+9. Once the `agent-config.json` file is populated and all corresponding ABI files are in the correct locations
+referred to in the `agent-config.json` file, the agent is complete.  Please test the agent against transactions
+that contain function calls that should trigger the agent.  Please also test the agent against transactions that should
```
Comment (arbitrary-sparrow) [inline gnosis-safe-multisig/SETUP.md:84]:
```suggestion
that contain emitted events that should trigger the agent.  Please also test the agent against transactions that should
```

File: gnosis-safe-multisig/package.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "forta-{developerAbbreviation}-{protocolAbbreviation}-dao-multisig",
```
Comment (arbitrary-sparrow) [inline gnosis-safe-multisig/package.json:2]:
```suggestion
  "name": "forta-{developerAbbreviation}-{protocolAbbreviation}-gnosis-safe-multisig",
```

File: gnosis-safe-multisig/SETUP.md (hunk 1 above)
Comment (echoAE) [reply gnosis-safe-multisig/SETUP.md]: Fixed.

File: gnosis-safe-multisig/SETUP.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+Note that any unused entries in the configuration file must be deleted for the agent to work.  The
+original version of the configuration file contains several placeholders to show the structure of
+the file, but these are not valid entries for running the agent.
+
+6. Create a new README.md file to provide a description of your agent, using examples from the Forta Github
+repository.  Also update the name and description fields in the `package.json` file.
+
+7. Move files to have the following directory structure:
+```
+  monitor-function-calls/
```
[the remark this answers is not in the document]
Comment (echoAE) [reply gnosis-safe-multisig/SETUP.md]: Fixed.

File: gnosis-safe-multisig/src/agent.js [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    data.provider = getEthersProvider();
+
+    data.previousBalances = {};
+
+    // get the current block number to retrieve all past Transfer events
+    const blockNumber = await data.provider.getBlockNumber();
+
+    // look up all Transfer events to this address
+    // eslint-disable-next-line import/no-dynamic-require,global-require
+    const { abi: erc20Abi } = require('../abi/ERC20.json');
```
[the remark this answers is not in the document]
Comment (echoAE) [reply gnosis-safe-multisig/src/agent.js:40]: Moved.

**Review by arbitrary-sparrow — APPROVED** [review]

**Review by arbitrary-sparrow — APPROVED** [review]

**Pull request merged**